src/FirebaseAuth.js

```diff
@@ -246,27 +245,41 @@
      * rejected if the client is unauthenticated and rejectIfAuthDataIsNull is true.
      */
     _routerMethodOnAuthPromise: function(rejectIfAuthDataIsNull) {
-      var ref = this._ref;
+      var ref = this._ref, utils = this._utils;
+      // wait for the initial auth state to resolve; on page load we have to request auth state
+      // asynchronously so we don't want to resolve router methods or flash the wrong state
+      return this._initialAuthResolver.then(function() {
+        // auth state may change in the future so rather than depend on the initially resolved state
+        // we also check the auth data (synchronously) if a new promise is requested, ensuring we resolve
+        // to the current auth state and not a stale/initial state (see https://github.com/firebase/angularfire/issues/590)
```

Comment:
I'd rather not reference GitHub issues from code comments. GitHub makes it trivial to look up this kind of information if we ever need to look at why code is the way it is.